sbin/atm/atm/atm.8
.\"
.\" ===================================
.\" HARP  |  Host ATM Research Platform
.\" ===================================
.\"
.\"
.\" This Host ATM Research Platform ("HARP") file (the "Software") is
.\" made available by Network Computing Services, Inc. ("NetworkCS")
.\" "AS IS".  NetworkCS does not provide maintenance, improvements or
.\" support of any kind.
.\"
.\" NETWORKCS MAKES NO WARRANTIES OR REPRESENTATIONS, EXPRESS OR IMPLIED,
.\" INCLUDING, BUT NOT LIMITED TO, IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS FOR A PARTICULAR PURPOSE, AS TO ANY ELEMENT OF THE
.\" SOFTWARE OR ANY SUPPORT PROVIDED IN CONNECTION WITH THIS SOFTWARE.
.\" In no event shall NetworkCS be responsible for any damages, including
.\" but not limited to consequential damages, arising from or relating to
.\" any use of the Software or related support.
.\"
.\" Copyright 1994-1998 Network Computing Services, Inc.
.\"
.\" Copies of this Software may be made, however, the above copyright
.\" notice must be reproduced on all copies.
.\"
.\" @(#) $FreeBSD: src/sbin/atm/atm/atm.8,v 1.3.2.2 2002/12/29 16:35:37 schweikh Exp $
.\" @(#) $DragonFly: src/sbin/atm/atm/atm.8,v 1.2 2003/06/17 04:27:32 dillon Exp $
.\"
.\"
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH ATM 8 "1998-08-20" "HARP"
.SH NAME
atm \- user configuration and display command for HARP ATM interface
.SH SYNOPSIS
Interface management subcommands:
.in +10
.ti -5
.B atm attach
<interface> <sigmgr>
.ti -5
.B atm detach
<interface>
.ti -5
.B atm set MAC
<interface> <MAC/ESI address>
.ti -5
.B atm set netif
<interface> <prefix> <count>
.ti -5
.B atm set prefix
<interface> <NSAP prefix>
.ti -5
.B atm show config
[<interface>]
.ti -5
.B atm show interface
[<interface>]
.ti -5
.B atm show netif
[<netif>]
.ti -5
.B atm show stats interface
[<interface> [phy | dev | atm | aal0 | aal4 | aal5 | driver]]
.sp
.ti -10
VCC management subcommands:
.ti -5
.B atm add PVC
<interface> <vpi> <vci> <aal> <encaps> <owner> ...
.ti -5
.B atm delete PVC
<interface> <vpi> <vci>
.ti -5
.B atm delete SVC
<interface> <vpi> <vci>
.ti -5
.B atm show stats VCC
[<interface> [<vpi> [<vci>]]]
.ti -5
.B atm show VCC
[<interface> [<vpi> [<vci>] | SVC | PVC]]
.sp
.ti -10
IP management subcommands:
.ti -5
.B atm add ARP
[<netif>] <host> <ATM address>
.ti -5
.B atm add PVC
<interface> <vpi> <vci> <aal> <encaps> IP <netif> <host> | dynamic
.ti -5
.B atm delete ARP
[<netif>] <host>
.ti -5
.B atm set arpserver
<netif> <ATM address> | local [<IP prefix> ...]
.ti -5
.B atm show ARP
[<host>]
.ti -5
.B atm show arpserver
[<netif>]
.ti -5
.B atm show IPVCC
[<host> | <netif>]
.ti -5
.sp
.ti -10
Miscellaneous subcommands:
.ti -5
.B atm help
.ti -5
.B atm show version
.in -10
.fi
.SH DESCRIPTION
.I atm
configures and displays the status of the Host ATM Research Platform
(HARP) networking software.
The subcommands fall into several categories:
.PP
\fIInterface management\fP subcommands allow manipulation of the
ATM interface.
Functions include assigning a signalling manager to an interface,
setting the ATM address, associating network interfaces with
an interface, and displaying information about interfaces.
.PP
\fIVCC management\fP subcommands allow for managing ATM virtual
channel connections (VCCs).
Functions include opening and closing VCCs and displaying information
about them.
.PP
\fIIP management\fP subcommands allow for managing the interface
between IP and the ATM software.
Functions include displaying and manipulating the ATMARP cache,
opening a PVC connected to IP,
assigning an ATMARP server to a network interface,
and displaying information about IP VCCs.
.PP
\fIMiscellaneous\fP subcommands allow for displaying the version
of the ATM software and for getting help with the \fIatm\fP command.
.SS "Signalling Managers"
The signalling manager is responsible for the opening and closing of
VCCs.
Four signalling managers are supported:
.PP
.in +10
.ti -5
PVC - for PVCs only,
.ti -5
SPANS - supports SPANS, FORE's proprietary signalling protocol,
.ti -5
UNI 3.0 - supports the signalling protocol from The ATM Forum's
\fIATM User-Network Interface Specification, Version 3.0\fP.
.ti -5
UNI 3.1 - supports the signalling protocol from The ATM Forum's
\fIATM User-Network Interface Specification, Version 3.1\fP.
.in -10
.PP
All four signalling managers support the opening and closing of PVCs
(see the \fIadd\fP and \fIdelete\fP subcommands).
.PP
A signalling manager must be attached to a physical interface
(see the \fIattach\fP subcommand)
before any VCCs can be created on the interface.
.SS "Physical and Network Interfaces"
Two types of interfaces are supported:
physical interfaces and network interfaces.
A physical interface represents a physical point of attachment to an
ATM network.
A physical interface has an ATM address associated with it, except
when the PVC-only signalling manager is being used.
.PP
A network interface is a logical interface.
One or more network interfaces are associated with a physical
interface;  each network interface has an IP address associated with it.
For UNI-controlled interfaces, there can be up to 256 network
interfaces associated with a physical interface.
In this case, the correspondence between the network interface and
the ATM address is determined by the selector field (the last
byte) of the physical interface's ATM address.
For PVC-only interfaces, there can be up to 256 logical interfaces
associated with each physical interface.
For interfaces controlled by the SPANS signalling manager,
there must be one and
only one network interface associated with each physical interface.
.SS "Keyword and Documentation Conventions"
Command and subcommand keywords can be abbreviated by simply giving
enough of the first part of the keyword to make it unique.
Thus, \fIatm sh v\fB gives the same result as \fIatm show vcc\fB.
.PP
All keywords are case-insensitive.
.PP
Where a host address needs to be given to the \fIatm\fP command,
either a DNS name or an IP address in dotted decimal format can
be used.
.PP
ATM addresses are specified as strings of hex digits, with an
optional leading "0x".
Fields within the address may be separated by periods, but periods
are for readability only and are ignored.
SPANS addresses are 8 bytes long, while NSAP-format addresses
are 20 bytes long.
The full address, including any leading zeroes, must be given.
For example:
.in +5
0x47.0005.80.ffe100.0000.f21a.0170.0020481a0170.00 (NSAP format)
.br
0x00000010.f2050aa9 (SPANS format)
.in -5
.fi
.SH SUBCOMMANDS
.SS Interface Management Subcommands:
.in +5
.ti -5
\fIatm add PVC <interface> <vpi> <vci> <aal> <encaps> <owner> ...\fP
.in -5
.PP
the format of the \fIadd PVC\fP subcommand varies depending on the
owner of the PVC.
See the description under "IP Management Subcommands."
.PP
\fIatm attach <interface> <sigmgr>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface to which the
signalling manager is to be attached,
.ti -5
\fI<sigmgr>\fP specifies which signalling manager is to be attached.
Valid choices are "SIGPVC", "SPANS", "UNI30", and "UNI31".
.in -10
.PP
This command attaches a signalling manager to an interface.
Until this is done, VCCs cannot be opened or closed.
Only one signalling manager at a time can be attached to an interface.
.PP
\fIatm detach <interface>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface whose signalling
manager is to be detached.
.in -10
.PP
This command detaches a signalling manager from an interface.
All VCCs that the signalling manager has created will be closed,
and no new VCCs can be created until a signalling manager (either
the same or a different one) is attached again.
.PP
\fIatm set MAC <interface> <MAC/ESI address>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface whose
MAC address is to be set,
.ti -5
\fI<MAC/ESI address>\fP specifies the 6-byte MAC part of the NSAP
address for the interface.
The MAC address is specified as a string of 12 hexadecimal
digits with an optional leading "0x".
Fields in the address may be separated by periods.
.in -10
.PP
This command sets the MAC address for a UNI-controlled interface.
The first 13 bytes (the prefix) of the 20-byte NSAP-format address
are set by the \fIatm set prefix\fP command or the ILMI daemon
(\fIilmid\fP (8)),
the next 6 bytes (the End System Identifier (ESI)) are set by
this command,
and the last byte (the selector) will be determined by which
network interface is to be associated with the address.
.PP
The \fIatm set MAC\fP command can be used to override the MAC
address in the interface hardware.
.PP
\fIatm set netif <interface> <prefix> <count>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface that the network
interface(s) are to be associated with,
.ti -5
\fI<prefix>\fP specifies the invariant part of the network
interface name,
.ti -5
\fI<count>\fP specifies the number of network interface to be
created.
.in -10
.PP
This command creates one or more network interfaces and associates them
with the specified physical interface.
The network interface names are determined by the prefix and the count.
The names will be of the form <prefix><nn>, where <prefix> is the
prefix specified in the \fIset\fP subcommand and <nn> is a number
in the range 0 - <count>-1.  For example, the command:
.PP
.ti +5
atm set netif hfa0 ni 2
.PP
would create two network interfaces, named ni0 and ni1, and associate
them with physical interface hfa0.
.PP
\fIatm set prefix <interface> <NSAP prefix>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface whose NSAP
prefix is to be set,
.ti -5
\fI<NSAP prefix>\fP specifies the first 13 bytes of the NSAP address
for the interface.
The prefix is specified as a string of hexadecimal digits with an
optional leading "0x".
Fields in the prefix may be separated by periods.
.in -10
.PP
This command sets the address for a UNI-controlled interface.
The first 13 bytes (the prefix) of the 20-byte NSAP-format address
are set by this command,
the next 6 bytes (the End System Identifier (ESI)) will be the
MAC address taken from the physical interface or set by the
\fIset MAC\fP subcommand,
and the last byte (the selector) will be determined by which
network interface is to be associated with the address.
.PP
The NSAP prefix must be set before a UNI-controlled
interface can become active.
This can be accomplished either by the ILMI daemon (\fIilmid\fP (8))
or the \fIset prefix\fP subcommand.
.PP
.I atm show config [<interface>]
.PP
displays the following information:
.PP
.B Interface
\- the name of the physical interface.
.PP
.B Vendor
\- the name of the adapter vendor.
.PP
.B Model
\- the model of the adapter.
.PP
.B Media
\- the communications medium used by the adapter.
.PP
.B Bus
\- the type of bus the adapter is attached to.
.PP
.B Serial No.
\- the adapter's serial number.
.PP
.B MAC address
\- the MAC address of the interface.
Note that this is the MAC address encoded in the hardware of
the adapter, even if the \fIatm set MAC\fP command has been used
to change the effective MAC address of the interface.
.PP
.B Hardware version
\- the hardware revision level reported by the interface.
.PP
.B Firmware version
\- the firmware revision level reported by the interface.
.PP
If no parameters are specified on the \fIshow config\fP subcommand,
the configurations of all physical interfaces will be displayed.
If an interface name is specified, only the configuration of the given
interface is displayed.
.PP
.I atm show interface [<interface>]
.PP
displays the following information:
.PP
.B Interface
\- the name of the physical interface.
.PP
.B Sigmgr
\- the name of the signalling manager which has been attached to the
interface.
A dash (-) is shown if no signalling manager has been attached.
.PP
.B State
\- the state of the signalling manager for the interface.
Each signalling manager has its own set of states.
They are:
.in +21
.ti -16
PVC:
.ti -11
ACTIVE\ ---\ The signalling manager is active.
.ti -11
DETACH\ ---\ The signalling manager is being detached.
.ti -16
SPANS:
.ti -11
ACTIVE\ ---\ The signalling manager is active.
.ti -11
DETACH\ ---\ The signalling manager is being detached.
.ti -11
INIT\ -----\ The signalling manager's initial state.
.ti -11
PROBE\ ----\ The signalling manager is attempting to make
contact with the ATM switch.
.ti -16
UNI 3.0 or UNI 3.1:
.ti -11
NULL\ -----\ The signalling manager's initial state.
.ti -11
ADR_WAIT\ -\ The signalling manager is waiting for the NSAP
prefix to be set.
.ti -11
INIT\ -----\ The signalling manager is attempting to establish
contact with the switch.
.ti -11
ACTIVE\ ---\ The signalling manager is active.
.ti -11
DETACH\ ---\ The signalling manager is being detached.
.ti -21
.PP
.B ATM address
\- the ATM address of the interface.
.PP
.B Network interfaces
\- the names of network interfaces, if any, associated with the
physical interface.
.PP
If no parameters are specified on the \fIshow interface\fP subcommand,
information about all physical interfaces will be displayed.
If an interface name is specified, only information about the given
interface is displayed.
.PP
.I atm show netif [<netif>]
.PP
displays the following information:
.PP
.B Net Intf
\- the name of the network interface.
.PP
.B IP Address
\- the IP address of the network interface.
.PP
If no parameters are specified on the \fIshow netif\fP subcommand,
information about all network interfaces will be displayed.
If an interface name is specified, only information about the given
network interface is displayed.
.PP
\fIatm show stats interface [<interface> [phy | dev | atm | aal0 |
aal4 | aal5 | driver]]\fP
.PP
displays statistics associated with one or more interfaces.
Subject-area keywords
(\fIphy\fP, \fIdev\fP, \fIatm\fP, \fIaal0\fP,
\fIaal4\fP, \fIaal5\fP, or \fIdriver\fP)
can be specified to change the scope of the statistics displayed.
.PP
If no subject area keyword is specified, the following information is
displayed:
.PP
.B Interface
\- the name of the physical ATM interface.
.PP
.B Input PDUs
\- the number of Protocol Data Units (PDUs) which have been received
by the interface.
.PP
.B Input Bytes
\- the number of bytes which have been received by the interface.
.PP
.B Input Errs
\- the number of input errors which the interface has experienced.
.PP
.B Output PDUs
\- the number of Protocol Data Units (PDUs) which have been transmitted
by the interface.
.PP
.B Output Bytes
\- the number of bytes which have been transmitted by the interface.
.PP
.B Output Errs
\- the number of output errors which the interface has experienced.
.PP
.B Cmd Errs
\- the number of command errors which the interface has experienced.
.PP
If a subject-area keyword is specified, then statistics for
that subject are displayed.
The statistics displayed depend on the adapter.
If requested statistics are not available for an adaptor,
an error will be noted.
.PP
If no parameters are specified on the \fIshow stats interface\fP
subcommand, statistics for all ATM interfaces are displayed.
If an interface name is specified, only statistics for the given
interface are displayed.
.PP
.SS VCC Management Subcommands:
.PP
\fIatm delete PVC <interface> <vpi> <vci>\fP
.br
\fIatm delete SVC <interface> <vpi> <vci>\fP
.PP
where:
.in +10
.ti -5
\fIPVC\fP specifies that the VCC to be closed is a PVC,
.ti -5
\fISVC\fP specifies that the VCC to be closed is an SVC,
.ti -5
\fI<interface>\fP specifies the physical interface at which the
VCC to be closed terminates,
.ti -5
\fI<vpi>\fP specifies the Virtual Path Identifier (VPI) of the VCC,
.ti -5
\fI<vci>\fP specifies the Virtual Channel Identifier (VCI) of the VCC.
.in -10
.PP
This command closes a VCC.
The two forms differ only in that the first specifies that the
VCC is a PVC (that was created by the \fIadd PVC\fP subcommand) and
the second specifies that the VCC is an SVC.
Reserved VCCs (with VCI values less than 32) cannot be closed
with this command.
.PP
\fIatm show stats VCC [<interface> [<vpi> [<vci>]]]\fP
.PP
displays the following information:
.PP
.B Interface
\- the physical interface on which the VCC terminates.
.PP
.B VPI
\- the Virtual Path Identifier (VPI) for the VCC.
.PP
.B VCI
\- the Virtual Channel Identifier (VCI) for the VCC.
.PP
.B Input PDUs
\- the number of Protocol Data Units (PDUs) which have been received
on the VCC.
.PP
.B Input Bytes
\- the number of bytes which have been received on the VCC.
.PP
.B Input Errs
\- the number of input errors which the VCC has experienced.
.PP
.B Output PDUs
\- the number of Protocol Data Units (PDUs) which have been transmitted
on the VCC.
.PP
.B Output Bytes
\- the number of bytes which have been transmitted on the VCC.
.PP
.B Output Errs
\- the number of output errors which the VCC has experienced.
.PP
If no parameters are specified on the \fIshow VCC\fP subcommand, all
active VCCs are displayed.
If an interface name is specified, all active VCCs for the given
interface are displayed.
If an interface and VPI are specified, all active VCCs for the VPI
on the given interface are displayed.
If an interface, VPI, and VCI are specified, only the specified VCC on
the given interface is displayed (note that this could actually be
two VCCs, since SPANS considers SVCs to be unidirectional).
.PP
\fIatm show VCC [<interface> [<vpi> [<vci>] | SVC | PVC]]\fP
.PP
displays the following information:
.PP
.B Interface
\- the physical interface on which the VCC terminates.
.PP
.B VPI
\- the Virtual Path Identifier (VPI) for the VCC.
.PP
.B VCI
\- the Virtual Channel Identifier (VCI) for the VCC.
.PP
.B AAL
\- the ATM Adaptation Layer (AAL) in use on the VCC.
Possible values are null and AAL 1-5.
.PP
.B Type
\- specifies whether the VCC is an SVC or a PVC.
.PP
.B Dir
\- the direction of information flow on the VCC.
VCCs can be inbound, outbound, or both.
.PP
.B State
\- the state of the VCC, as reported by the signalling manager.
Each signalling manager has its own set of states.
They are:
.in +21
.ti -16
PVC:
.ti -11
NULL\ -----\ No state.
.ti -11
ACTIVE\ ---\ The VCC is active.
.ti -11
FREE\ -----\ The VCC is closed and the signalling manager is waiting for
its resources to be freed.
.ti -16
SPANS:
.ti -11
NULL\ -----\ No state.
.ti -11
ACTIVE\ ---\ The VCC is a PVC and is active.
.ti -11
ACT_DOWN\ -\ The VCC is a PVC and the interface is down.
.ti -11
POPEN\ ----\ The VCC is being opened.
.ti -11
R_POPEN\ --\ The VCC is being opened by a remote host.
.ti -11
OPEN\ -----\ The VCC is active.
.ti -11
CLOSE\ ----\ The VCC is being closed.
.ti -11
ABORT\ ----\ The VCC is being aborted.
.ti -11
FREE\ -----\ The VCC is closed and the signalling manager is waiting for
its resources to be freed.
.ti -16
UNI 3.0 or UNI 3.1:
.ti -11
NULL\ -----\ No state.
.ti -11
C_INIT\ ---\ A VCC is being initiated.
.ti -11
C_OUT_PR\ -\ An outgoing VCC request is proceeding.
.ti -11
C_PRES\ ---\ A VCC is being initiated by the network.
.ti -11
CONN_REQ\ -\ A VCC request has been accepted by a HARP user.
.ti -11
C_IN_PR\ --\ An incoming VCC request is proceeding.
.ti -11
ACTIVE\ ---\ The VCC is active.
.ti -11
REL_REQ\ --\ The VCC is being closed.
.ti -11
REL_IND\ --\ The network is clearing a VCC.
.ti -11
SSCF_REC\ -\ The SSCF session on the signalling channel is in
recovery from an error.
.ti -11
FREE\ -----\ The VCC is closed and the signalling manager is waiting
for its resources to be freed.
.ti -11
ACT_DOWN\ -\ The VCC is a PVC and the interface is down.
.ti -21
.PP
.B Encaps
\- the encapsulation in effect on the VCC.
Possible encapsulations are null and LLC/SNAP.
.PP
.B Owner
\- the owner or owners of the VCC.
Shows the name(s) of the function(s) using the VCC.
.PP
.B Destination
\- the ATM address of the host at the remote end of the VCC.
.PP
If no parameters are specified on the \fIshow VCC\fP subcommand, all
active VCCs are displayed.
If an interface name is specified, all active VCCs for the given
interface are displayed.
If an interface and VPI are specified, all active VCCs for the VPI
on the given interface are displayed.
If an interface, VPI, and VCI are specified, only the specified VCC on
the given interface is displayed (note that this could actually be
two VCCs, since SPANS considers SVCs to be unidirectional).
.PP
.SS IP Management Subcommands:
\fIatm add ARP [<netif>] <host> <ATM address>\fP
.PP
where:
.in +10
.ti -5
\fI<netif>\fP is the optional name of the network interface the
ATMARP entry is to be associated with.
If no name is specified, a network interface is chosen depending
on the IP address of the host being added.
.ti -5
\fI<host>\fP is the host name or IP address of the host to
be added to the ATMARP table,
.ti -5
\fI<ATM address>\fP is the ATM address of the host.
.in -10
.PP
This command adds an entry to the ATMARP table for ATM.
The given host's IP address is associated with the given ATM address.
When IP needs to transmit data to the host, the specified ATM
address will be used to open an SVC.
.PP
The entry will be marked as permanent in the ATMARP table and will not
be subject to aging.
.PP
.in +5
.ti -5
\fIatm add PVC <interface> <vpi> <vci> <aal> <encaps> IP <netif> <host> | dynamic\fP
.in -5
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface where the PVC
is to terminate,
.ti -5
\fI<vpi>\fP specifies the Virtual Path Identifier (VPI) of the PVC,
.ti -5
\fI<vci>\fP specifies the Virtual Channel Identifier (VCI) of the PVC,
.ti -5
\fI<aal>\fP specifies the ATM Adaptation Layer (AAL) for the PVC.
Valid choices are "null" or "AAL0" for the null AAL; "AAL1" for
AAL 1; "AAL2" for AAL 2; "AAL3", "AAL4", or "AAL3/4" for AAL 3/4;
and "AAL5" for AAL 5,
.ti -5
\fI<encaps>\fP specifies the encapsulation for the PVC.
Valid choices are "null" or "none" for null encapsulation, and
"LLC/SNAP", "LLC", or "SNAP" for LLC/SNAP encapsulation,
.ti -5
\fIIP\fP specifies that the owner of the PVC is IP.
.ti -5
\fI<netif>\fP specifies the network interface which the PVC is
to be associated with.
The network interface must exist and be associated with the
specified physical interface,
.ti -5
\fI<host> | dynamic\fP gives the address of the host at
the far end of the PVC, or the word "dynamic" if its address
is to be determined with Inverse ARP.
If "dynamic" is specified, LLC/SNAP encapsulation must also
be specified.
.PP
This command creates a PVC with the specified attributes and attaches
it to IP.
.PP
\fIatm delete ARP [<netif>] <host>\fP
.PP
where:
.in +10
.ti -5
\fI<netif>\fP is the optional name of the network interface the
ATMARP entry is associated with.
If no name is specified, the specified host is deleted from the
cache regardless of what network interface it is associated with.
.ti -5
\fI<host>\fP is the host name or IP address of the host to
be deleted from the ATMARP table.
.PP
This command deletes the specified host's entry from the ATMARP table.
.PP
\fIatm set arpserver <netif> <ATM address> | local [<IP prefix> ...]\fP
.PP
where:
.in +10
.ti -5
\fI<netif>\fP specifies the network interface for which the
ATMARP server address is to be set.
.ti -5
\fI<ATM address>\fP specifies the ATM address of the host which is to
provide ATMARP service.
If "local" is specified instead of an ATM address, the host on
which the command is issued will become the ATMARP server.
.ti -5
\fI<IP prefix> ...\fP is an optional list of IP prefixes
that the ATMARP server will provide information about.
An IP prefix is specified as a dotted decimal IP address, followed by
a slash, followed a number specifying how many bits of the IP address
are significant.
For example, 10.0.0.0/8 indicates that the ATMARP server will provide
services for all addresses on IP network 10.
The IP subnetwork which the network interface belongs to is
automatically included.
.in -10
.PP
This command sets the address of the ATMARP server for a network
interface.
.PP
.I atm show ARP [<host>]
.PP
displays the following information:
.PP
.B Net Intf
\- the network interface which traffic for the entry will use.
.PP
.B Flags
\- flags showing whether the entry is valid and whether it is
permanent.
\- flags giving further information about the ATMARP entry.
The meanings of the characters in the flags are:
.PP
.in +5
P - the entry is permanent
.br
R - the entry has been refreshed
.br
V - the entry is valid
.in -5
.PP
.B Age
\- the number of minutes for which the entry will remain valid.
.PP
.B Origin
\- the source of the ATMARP entry.
Possible values are:
.in +16
.ti -11
LOCAL\ ----\ The entry is for an interface on the host.
.ti -11
PERM\ -----\ The entry is permanent.
This is used for entries that are created with the
\fIadd ARP\fP command.
.ti -11
REG\ ------\ The entry was created as the result of a host
registering with the ATMARP server.
.ti -11
SCSP\ -----\ The entry was learned via SCSP.
.ti -11
LOOKUP\ ---\ The entry was created as the result of a host
performing an ATMARP lookup.
.ti -11
PEER_RSP\ -\ The entry was created as the result of a host
answering an InARP Request.
.ti -11
PEER_REQ\ -\ The entry was created as the result of a host
sending an InARP Request.
.in -5
.PP
.B ATM address
\- the ATM address of the host the entry refers to.
.PP
.B IP address
\- the IP address or domain name of the host the entry refers to.
.PP
If no parameters are specified on the \fIshow ARP\fP subcommand,
the whole ATMARP table will be displayed.
If a host name or IP address is specified, only information about the
given host is displayed.
.PP
This command displays both information that has been learned dynamically
(through one form or another of ATMARP and via SCSP) and information
which has been configured by the user (through the \fIadd ARP\fP
subcommand).
.PP
.I atm show arpserver [<netif>]
.PP
displays the following information:
.PP
.B Net Intf
\- the network interface for which information is being displayed.
.PP
.B State
\- the state of the connection to the ATMARP server.
Possible values are:
.in +16
.ti -11

usr.sbin/ppp/atm.c
/*-
 * Copyright (c) 2000 Jakob Stoklund Olesen <stoklund@taxidriver.dk>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD: src/usr.sbin/ppp/atm.c,v 1.9.2.1 2002/09/01 02:12:22 brian Exp $
 * $DragonFly: src/usr.sbin/ppp/atm.c,v 1.2 2003/06/17 04:30:00 dillon Exp $
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <net/if.h>
#include <netnatm/natm.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <sys/uio.h>
#include <termios.h>
#include <unistd.h>

#include "layer.h"
#include "defs.h"
#include "mbuf.h"
#include "log.h"
#include "timer.h"
#include "lqr.h"
#include "hdlc.h"
#include "throughput.h"
#include "fsm.h"
#include "lcp.h"
#include "ccp.h"
#include "link.h"
#include "async.h"
#include "descriptor.h"
#include "physical.h"
#include "main.h"
#include "atm.h"

/* String identifying PPPoA */
#define PPPOA		"PPPoA"
#define PPPOA_LEN	(sizeof(PPPOA) - 1)

struct atmdevice {
  struct device dev;		/* What struct physical knows about */
};

#define device2atm(d) ((d)->type == ATM_DEVICE ? (struct atmdevice *)d : NULL)

int
atm_DeviceSize(void)
{
  return sizeof(struct atmdevice);
}

static ssize_t
atm_Sendto(struct physical *p, const void *v, size_t n)
{
  ssize_t ret = write(p->fd, v, n);
  if (ret < 0) {
    log_Printf(LogDEBUG, "atm_Sendto(%ld): %s\n", (long)n, strerror(errno));
    return ret;
  }
  return ret;
}

static ssize_t
atm_Recvfrom(struct physical *p, void *v, size_t n)
{
    ssize_t ret = read(p->fd, (char*)v, n);
    if (ret < 0) {
      log_Printf(LogDEBUG, "atm_Recvfrom(%ld): %s\n", (long)n, strerror(errno));
      return ret;
    }
    return ret;
}

static void
atm_Free(struct physical *p)
{
  struct atmdevice *dev = device2atm(p->handler);

  free(dev);
}

static void
atm_device2iov(struct device *d, struct iovec *iov, int *niov,
               int maxiov, int *auxfd, int *nauxfd)
{
  int sz = physical_MaxDeviceSize();

  iov[*niov].iov_base = realloc(d, sz);
  if (iov[*niov].iov_base == NULL) {
    log_Printf(LogALERT, "Failed to allocate memory: %d\n", sz);
    AbortProgram(EX_OSERR);
  }
  iov[*niov].iov_len = sz;
  (*niov)++;
}

static const struct device baseatmdevice = {
  ATM_DEVICE,
  "atm",
  0,
  { CD_NOTREQUIRED, 0 },
  NULL,
  NULL,
  NULL,
  NULL,
  NULL,
  NULL,
  NULL,
  atm_Free,
  atm_Recvfrom,
  atm_Sendto,
  atm_device2iov,
  NULL,
  NULL,
  NULL
};

struct device *
atm_iov2device(int type, struct physical *p, struct iovec *iov, int *niov,
               int maxiov, int *auxfd, int *nauxfd)
{
  if (type == ATM_DEVICE) {
    struct atmdevice *dev = (struct atmdevice *)iov[(*niov)++].iov_base;

    dev = realloc(dev, sizeof *dev);	/* Reduce to the correct size */
    if (dev == NULL) {
      log_Printf(LogALERT, "Failed to allocate memory: %d\n",
                 (int)(sizeof *dev));
      AbortProgram(EX_OSERR);
    }

    /* Refresh function pointers etc */
    memcpy(&dev->dev, &baseatmdevice, sizeof dev->dev);

    physical_SetupStack(p, dev->dev.name, PHYSICAL_FORCE_SYNCNOACF);
    return &dev->dev;
  }

  return NULL;
}

static struct atmdevice *
atm_CreateDevice(struct physical *p, const char *iface, unsigned vpi,
                 unsigned vci)
{
  struct atmdevice *dev;
  struct sockaddr_natm sock;

  if ((dev = calloc(1, sizeof *dev)) == NULL) {
    log_Printf(LogWARN, "%s: Cannot allocate an atm device: %s\n",
               p->link.name, strerror(errno));
    return NULL;
  }

  sock.snatm_len = sizeof sock;
  sock.snatm_family = AF_NATM;
  strncpy(sock.snatm_if, iface, IFNAMSIZ);
  sock.snatm_vpi = vpi;
  sock.snatm_vci = vci;

  log_Printf(LogPHASE, "%s: Connecting to %s:%u.%u\n", p->link.name,
             iface, vpi, vci);

  p->fd = socket(PF_NATM, SOCK_DGRAM, PROTO_NATMAAL5);
  if (p->fd >= 0) {
    log_Printf(LogDEBUG, "%s: Opened atm socket %s\n", p->link.name,
               p->name.full);
    if (connect(p->fd, (struct sockaddr *)&sock, sizeof sock) == 0)
      return dev;
    else
      log_Printf(LogWARN, "%s: connect: %s\n", p->name.full, strerror(errno));
  } else
    log_Printf(LogWARN, "%s: socket: %s\n", p->name.full, strerror(errno));

  close(p->fd);
  p->fd = -1;
  free(dev);

  return NULL;
}

struct device *
atm_Create(struct physical *p)
{
  struct atmdevice *dev;

  dev = NULL;
  if (p->fd < 0 && !strncasecmp(p->name.full, PPPOA, PPPOA_LEN)
      && p->name.full[PPPOA_LEN] == ':') {
    char iface[25];
    unsigned vci, vpi;

    if (sscanf(p->name.full + PPPOA_LEN + 1, "%25[A-Za-z0-9]:%u.%u", iface,
               &vpi, &vci) != 3) {
      log_Printf(LogWARN, "Malformed ATM device name \'%s\', "
                 "PPPoA:if:vpi.vci expected\n", p->name.full);
      return NULL;
    }

    dev = atm_CreateDevice(p, iface, vpi, vci);
  }

  if (dev) {
    memcpy(&dev->dev, &baseatmdevice, sizeof dev->dev);
    physical_SetupStack(p, dev->dev.name, PHYSICAL_FORCE_SYNCNOACF);
    if (p->cfg.cd.necessity != CD_DEFAULT)
      log_Printf(LogWARN, "Carrier settings ignored\n");
    return &dev->dev;
  }

  return NULL;
}
